python/context_selector/context_widget.py

```diff
@@ -303,28 +303,96 @@ def set_up(self, task_manager):
         task_manager.task_completed.connect(self._on_task_completed)
         task_manager.task_failed.connect(self._on_task_failed)
 
-        # Query Shotgun for valid entity types for PublishedFile.entity field
-        self._schema_query_id = task_manager.add_task(
-            _query_publishedfile_entity_schema)
-
         # query all my assigned tasks in a bg task
         self._my_tasks_query_id = task_manager.add_task(_query_my_tasks)
 
         # get recent contexts from user settings
         self._get_recent_contexts()
 
+    def restrict_entity_types_by_link(self, entity_type, field_name):
+        """
+        Specify what entries should show up in the list of links when
+        using the auto completer.
+
+        For the simple case where you just want to show a given set of
+        entity types, use :meth:`restrict_entity_types`. This method is
+        a more complex restriction suitable for workflows around publishing
+        and review.
+
+        This method will look at the given link field (e.g. ``PublishedFile.entity``)
+        and inspect the shotgun schema to see which entity types are valid connections
+        to this field (e.g. in this example which entity types can you can associate
+        a publish with) and those types will appear in the list of items shown by the
+        auto completer.
+
+        This is useful when you want to use the context widget in conjunction with
+        workflows related to for example publishes, versions or notes and you want to
+        restrict the entities displayed by the auto completer to the ones that have been
+        configured in the shotgun site schema to be able to associate with the given type.
+
+        :param str entity_type: Entity type to restrict based on
+        :param str field_name: Shotgun field to restrict based on
+        """
+        if self._task_manager is None:
+            raise RuntimeError("You must run set_up() before this method can be executed.")
+
+        # Query Shotgun for valid entity types for PublishedFile.entity field
+        self._schema_query_id = self._task_manager.add_task(
+            _query_publishedfile_entity_schema,
+            task_args=[entity_type, field_name]
+        )
+
+    def restrict_entity_types(self, entity_types):
+        """
+        Restrict which entity types should show up in the list of matches.
+
+        :param list entity_types: List of entity types
+        """
+        logger.debug(
+            "Restricting auto completer to show the following types: %s" % entity_types
+        )
+        # construct a dictionary that the search widget expects for
+        # filtering. This is a dictionary with the entity types as keys and
+        # values a list of search filters. We don't have any filters, so we
```

Comment:
worth allowing the calling code to supply filters here?

Reply:
that syntax is a bit iffy IIRC - i'd rather leave that out for the time being - we can always add it in later! Good point though :)